File: sdk/node/post.go
package node

import (
	"bytes"

	"github.com/bpprotocol/blockparty/sdk/block"
	"github.com/bpprotocol/blockparty/sdk/blockpb"
	"github.com/bpprotocol/blockparty/sdk/blocktypes"
	"github.com/bpprotocol/blockparty/sdk/derive"
	"github.com/bpprotocol/blockparty/sdk/encryption"
	"github.com/bpprotocol/blockparty/sdk/identity"
)

// ContentTypeHeader is the required MIME header key for content payloads.
const ContentTypeHeader = "bpprotocol.org/v1/content-type"

// BuildPost creates a signed content.post block whose text body is encrypted to
// an audience (identified by audienceCode + audienceSecret) within a World.
func BuildPost(w derive.World, author identity.Identity, audienceCode derive.Code, audienceSecret []byte, timestamp int64, text string) (*blockpb.Block, error) {
	typeCode := derive.GetTypeCode(w, blocktypes.TypeContentPost)
	payload, err := blocktypes.MarshalPayload(&blockpb.ContentPost{
		Headers: map[string]string{ContentTypeHeader: "text/plain"},
		Body:    []byte(text),
	})
	if err != nil {
		return nil, err
	}
	data, err := encryption.EncryptForBlock(audienceSecret, block.Version, []byte(typeCode), []byte(audienceCode), timestamp, payload)
	if err != nil {
		return nil, err
	}
	b := block.New(author.Address, typeCode, audienceCode, timestamp, data)
	block.Sign(b, w, author.MLDSA)
	return b, nil
}

// OpenPost decrypts and decodes a content.post block with a known audience
// secret, returning the text body. ok is false if the block is not a
// content.post for this World. Decryption is authenticated (AEAD), so a
// tampered block yields an error.
func OpenPost(w derive.World, r *blocktypes.Resolver, b *blockpb.Block, audienceSecret []byte, text *string) (ok bool, err error) {
	urn, found := r.TypeURN(b.TypeCode)
	if !found || urn != blocktypes.TypeContentPost {
		return false, nil
	}
	plaintext, err := encryption.DecryptData(audienceSecret, b)
	if err != nil {
		return false, err
	}
	msg, err := blocktypes.DecodePayload(urn, plaintext)
	if err != nil {
		return false, err
	}
	post, ok := msg.(*blockpb.ContentPost)
	if !ok {
		return false, nil
	}
	*text = string(bytes.Clone(post.Body))
	return true, nil
}

File: js/packages/cli/src/post.ts
import {
  create,
  toBinary,
  utf8ToBytes,
  getTypeCode,
  newBlock,
  signBlock,
  encryptForBlock,
  decryptData,
  decodePayload,
  ContentPostSchema,
  TYPE_CONTENT_POST,
  type World,
  type Identity,
  type Resolver,
  type Block,
  type ContentPost,
} from "@blockparty/sdk";

const CONTENT_TYPE_HEADER = "bpprotocol.org/v1/content-type";
const decoder = new TextDecoder();

/**
 * Build a signed content.post block whose text body is encrypted to an audience
 * (identified by code + secret) within a World.
 */
export function buildPost(
  world: World,
  author: Identity,
  audienceCode: Uint8Array,
  audienceSecret: Uint8Array,
  timestamp: number,
  text: string,
): Block {
  const typeCode = getTypeCode(world, TYPE_CONTENT_POST);
  const payload = toBinary(
    ContentPostSchema,
    create(ContentPostSchema, {
      headers: { [CONTENT_TYPE_HEADER]: "text/plain" },
      body: utf8ToBytes(text),
    }),
  );
  const data = encryptForBlock(audienceSecret, 1, typeCode, audienceCode, timestamp, payload);
  const block = newBlock(author.address, typeCode, audienceCode, timestamp, data);
  signBlock(block, world, author.mldsa);
  return block;
}

/**
 * Decrypt and decode a content.post block with a known audience secret,
 * returning the text body, or undefined if it is not a content.post for this
 * World. Decryption is authenticated, so a tampered block throws.
 */
export function openPost(
  resolver: Resolver,
  block: Block,
  audienceSecret: Uint8Array,
): string | undefined {
  if (resolver.typeURN(block.typeCode) !== TYPE_CONTENT_POST) return undefined;
  const plaintext = decryptData(audienceSecret, block);
  const post = decodePayload(TYPE_CONTENT_POST, plaintext) as ContentPost;
  return decoder.decode(post.body);
}
